{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1ecf1248-03a7-4d3a-bcff-6faee8355d82",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib.cm as cm\n",
    "# from google.colab import drive\n",
    "import matplotlib.lines as mlines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b1405d34-3b1b-4d49-a3b1-c3dd85b94abb",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({\n",
    "    \"text.usetex\": False,\n",
    "    \"font.family\": \"sans-serif\",\n",
    "    \"font.sans-serif\": \"Helvetica\",\n",
    "    \"font.weight\": \"normal\",\n",
    "    \"font.size\": 22\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ba70b24a-f28c-4e75-9d30-8c4dfc4f3b0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def defparse(astr):\n",
    "    dd = {}\n",
    "    y = astr.strip().split(';')\n",
    "    for a in y:\n",
    "        b = a.split('=')\n",
    "        dd[b[0]] = b[1]\n",
    "    return dd\n",
    "\n",
    "# Define the gff_parse function to parse the GFF3 file and extract gene models\n",
    "def gff_parse(afile, tchr, rstart, rstop):\n",
    "    fh = open(afile)\n",
    "    gene_dict = {}\n",
    "    for x in fh:\n",
    "        if x[0] == '#': continue\n",
    "        y = x.strip().split('\\t')\n",
    "        if y[0] != f\"{tchr}\": continue\n",
    "        if int(y[3]) > rstop + 100000: continue\n",
    "        if int(y[4]) < rstart - 100000: continue\n",
    "        defdict = defparse(y[-1])\n",
    "        if y[2] == 'mRNA' and defdict.get(\"longest\") == '1':\n",
    "            mygene = defdict[\"ID\"]\n",
    "            gene_dict[mygene] = {\"start\": int(y[3]), \"stop\": int(y[4]), \"utr\": [], \"cds\": [], \"strand\": y[6]}\n",
    "        if y[2] == 'CDS':\n",
    "            mygene = defdict[\"Parent\"]\n",
    "            if not mygene in gene_dict: continue\n",
    "            gene_dict[mygene][\"cds\"].append((int(y[3]), int(y[4]) - int(y[3])))\n",
    "        if y[2] == \"five_prime_UTR\" or y[2] == \"three_prime_UTR\":\n",
    "            mygene = defdict[\"Parent\"]\n",
    "            if not mygene in gene_dict: continue\n",
    "            gene_dict[mygene][\"utr\"].append((int(y[3]), int(y[4]) - int(y[3])))\n",
    "    fh.close()\n",
    "    return gene_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8fd2588d-ccc9-43ea-ba6d-3fcbf4b56c44",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'plt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 6\u001b[0m\n\u001b[1;32m      4\u001b[0m windowsize \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m35000\u001b[39m  \u001b[38;5;66;03m# Adjusted to ensure all genes are visible\u001b[39;00m\n\u001b[1;32m      5\u001b[0m stepsize \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m15000\u001b[39m  \u001b[38;5;66;03m# Adjusted for better labeling\u001b[39;00m\n\u001b[0;32m----> 6\u001b[0m plt\u001b[38;5;241m.\u001b[39mrcParams\u001b[38;5;241m.\u001b[39mupdate({\n\u001b[1;32m      7\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtext.usetex\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m      8\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfont.family\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msans-serif\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m      9\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfont.sans-serif\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mHelvetica\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     10\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfont.weight\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnormal\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     11\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfont.size\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;241m22\u001b[39m\n\u001b[1;32m     12\u001b[0m })\n\u001b[1;32m     13\u001b[0m \u001b[38;5;66;03m# Load gene positions\u001b[39;00m\n\u001b[1;32m     14\u001b[0m gene_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mZmays_833_Zm-B73-REFERENCE-NAM-5.0.55.gene.gff3\u001b[39m\u001b[38;5;124m'\u001b[39m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'plt' is not defined"
     ]
    }
   ],
   "source": [
    "target_chr = 9\n",
    "target_loc = 21785915\n",
    "additional_target_loc =21755970# New target locus to highlight\n",
    "windowsize = 35000  # Adjusted to ensure all genes are visible\n",
    "stepsize = 15000  # Adjusted for better labeling\n",
    "plt.rcParams.update({\n",
    "    \"text.usetex\": True,\n",
    "    \"font.family\": \"sans-serif\",\n",
    "    \"font.sans-serif\": \"Helvetica\",\n",
    "    \"font.weight\": \"normal\",\n",
    "    \"font.size\": 22\n",
    "})\n",
    "# Load gene positions\n",
    "gene_file = 'Zmays_833_Zm-B73-REFERENCE-NAM-5.0.55.gene.gff3'\n",
    "gene_dict = gff_parse(gene_file, target_chr, target_loc - windowsize, target_loc + windowsize)\n",
    "# Load GWAS hits\n",
    "gwas_file = 'eQTL_significant_SNPs_Zm00001eb33170.csv'\n",
    "eQTL_hits = pd.read_csv(gwas_file)\n",
    "eQTL_hits = eQTL_hits.loc[eQTL_hits['CHROM'] == target_chr]\n",
    "# Load RMIP values\n",
    "rmip_file = 'multispeqcutoff_final.csv'\n",
    "rmip_data = pd.read_csv(rmip_file)\n",
    "rmip_data = rmip_data.loc[rmip_data['CHROM'] == target_chr]\n",
    "# Load LD values\n",
    "ld_file = 'ld_chr9.csv'\n",
    "ld_data = pd.read_csv(ld_file)\n",
    "# # Merge LD values with eQTL hits, using 'R2' as the LD measure\n",
    "eQTL_hits = pd.merge(eQTL_hits, ld_data[['POS', 'R2']], on='POS', how='left')\n",
    "# # Exclude rows with missing R2 values\n",
    "eQTL_hits = eQTL_hits.dropna(subset=['R2'])\n",
    "cmap = cm.get_cmap('Blues')\n",
    "norm = mcolors.Normalize(vmin=0, vmax=1)\n",
    "# Set up the figure with desired size (1200px width, 710px height)\n",
    "fig, ax2 = plt.subplots(figsize=(17,6))\n",
    "# Set up the secondary axis (ax1) for log10(p-value)\n",
    "ax1 = ax2.twinx()\n",
    "# Add significance threshold on the secondary axis (log10(p-value))\n",
    "ax1.plot([target_loc - windowsize, target_loc + windowsize], [7, 7], '--', color='black', linewidth=2.5)\n",
    "# Plot gene models on the secondary axis (log10(p-value))\n",
    "for g in gene_dict:\n",
    "    myname = g.split('_')[0]\n",
    "    if not myname in ['Zm00001eb377160', 'Zm00001eb377130', 'Zm00001eb377150', 'Zm00001eb377140']: continue\n",
    "    if gene_dict[g][\"strand\"] == '+':\n",
    "        myoffset = 3\n",
    "        arrow = mpatches.FancyArrowPatch((gene_dict[g][\"start\"], myoffset+0.75), (gene_dict[g][\"stop\"] + 100, myoffset+0.75),\n",
    "                                         arrowstyle='-|>', mutation_scale=10, color=\"black\")\n",
    "        color = 'red' if myname == 'Zm00001eb377130' else 'black'\n",
    "        ax1.text((gene_dict[g][\"start\"] + gene_dict[g][\"stop\"]) / 2, 4.575, myname, ha=\"center\", color=color, fontsize=20)\n",
    "    if gene_dict[g][\"strand\"] == '-':\n",
    "        myoffset = 1\n",
    "        arrow = mpatches.FancyArrowPatch((gene_dict[g][\"stop\"], myoffset+0.75), (gene_dict[g][\"start\"] - 100, myoffset+0.75),\n",
    "                                         arrowstyle='-|>', mutation_scale=10, color=\"black\")\n",
    "        color = 'red' if myname == 'Zm00001eb377130' else 'black'\n",
    "        ax1.text((gene_dict[g][\"start\"] + gene_dict[g][\"stop\"]) / 2, 0.0, myname, ha=\"center\", color=color, fontsize=20)\n",
    "    ax1.broken_barh(gene_dict[g][\"utr\"], (myoffset, 1.5), color=\"blue\")\n",
    "    ax1.broken_barh(gene_dict[g][\"cds\"], (myoffset, 1.5), color=\"green\")\n",
    "    ax1.add_patch(arrow)\n",
    "    if (gene_dict[g][\"start\"] + gene_dict[g][\"stop\"]) / 2 < target_loc - windowsize or (gene_dict[g][\"start\"] + gene_dict[g][\"stop\"]) / 2 > target_loc + windowsize: continue\n",
    "    if myname in ['Zm00001eb377160', 'Zm00001eb377130', 'Zm00001eb377150']:\n",
    "        color = 'red' if myname == 'Zm00001eb377130' else 'black'\n",
    "\n",
    "# Plot GWAS results with LD values on the secondary axis (log10(p-value))\n",
    "sc = ax1.scatter(eQTL_hits[\"POS\"], eQTL_hits[\"logpvalue\"], c=eQTL_hits[\"R2\"], cmap=cmap, norm=norm, label=\"GWAS Peaks\", s=70)\n",
    "\n",
    "# Plot RMIP values on the primary axis (left, ax2)\n",
    "ax2.set_ylabel('RMIP (Relative Chlorophyll)', color=\"black\")\n",
    "ax2.set_ylim([0, 0.3])\n",
    "ax2.scatter(rmip_data[\"POS\"], rmip_data[\"support\"], color=\"orange\", label=\"RMIP Values\")\n",
    "\n",
    "# # Highlight the additional target locus with a triangle based on RMIP dataset\n",
    "highlighted_rmip_value = rmip_data.loc[rmip_data['POS'] == additional_target_loc, 'support'].values[0]\n",
    "ax2.scatter(additional_target_loc, highlighted_rmip_value, color='darkred', marker='^', s=200, label=\"Highlighted Hit\")\n",
    "\n",
    "# # Add proper x-axis labels\n",
    "myxticks = list(range(round(target_loc - windowsize, -4), round(target_loc + windowsize, -4) + stepsize, stepsize))\n",
    "myxticklabels = [\"{:,} MB\".format(round(t / 1000000,2)) for t in myxticks]\n",
    "\n",
    "# # Apply visual tweaks\n",
    "ax2.spines['top'].set_visible(False)\n",
    "ax2.spines['right'].set_visible(False)\n",
    "ax1.spines['top'].set_visible(False)\n",
    "for axis in ['bottom', 'left']:\n",
    "    ax2.spines[axis].set_linewidth(2.8)  # Set axis line thickness to 1.5\n",
    "ax1.spines['right'].set_linewidth(2.8)  # Set axis line thickness to 1.5\n",
    "\n",
    "# ax2.spines['left'].set_linewidth(1.5)  # Set axis line thickness to 1.5\n",
    "ax2.set_xticks(myxticks)\n",
    "ax2.set_xticklabels(myxticklabels,color=\"black\")\n",
    "ax2.set_xlabel(f\"Position on Chromosome {target_chr}\", color=\"black\")\n",
    "ax2.set_xlim([target_loc - windowsize, target_loc + windowsize])\n",
    "ax1.set_ylim([-0.1, 30])\n",
    "ax1.set_ylabel(\"log10(p-value) (eQTL)\", color=\"black\")\n",
    "\n",
    "ax2.tick_params(width=2, length=10, labelsize=22)  # Set tick label size\n",
    "ax1.tick_params(width=2, length=10, labelsize=22)  # Set tick label size\n",
    "\n",
    "# Add colorbar for LD values (R2) at the top center\n",
    "cbar = plt.colorbar(sc, ax=[ax1, ax2], orientation='horizontal', pad=-0.1, shrink=0.2, aspect=15)\n",
    "# Set the label with only \"2\" as a superscript\n",
    "cbar.set_label('RÂ²', fontsize=22)\n",
    "cbar.set_ticks([0, 0.2, 0.4, 0.6, 0.8, 1])\n",
    "cbar.set_ticklabels(['0', '0.2', '0.4', '0.6', '0.8', '1'])\n",
    "yticks1= ax2.get_yticks()\n",
    "yticks2= ax1.get_yticks()\n",
    "ax2.set_yticklabels([round(x,3) for x in yticks1])\n",
    "ax1.set_yticklabels([round(x) for x in yticks2])\n",
    "\n",
    "custom_yticks_primary = [0.0, 0.1, 0.2, 0.3]  # Your custom y-tick values\n",
    "ax2.set_yticks(custom_yticks_primary)\n",
    "ax2.set_yticklabels([f\"{x:.2f}\" for x in custom_yticks_primary])\n",
    "\n",
    "custom_yticks_secondary = [0, 7, 14, 21]  # Your custom y-tick values\n",
    "ax1.set_yticks(custom_yticks_secondary)\n",
    "ax1.set_yticklabels([f\"{x:.0f}\" for x in custom_yticks_secondary])\n",
    "\n",
    "# Define custom legend handles for RMIP and eQTL SNPs\n",
    "rmip_handle = mlines.Line2D([], [], color='darkred', marker='^', linestyle='None', markersize=15, label='RMIP Highlighted Hit')\n",
    "eqtl_handle = mlines.Line2D([], [], color='darkblue', marker='o', linestyle='None', markersize=10, label='eQTL SNPs')\n",
    "\n",
    "# Add custom legend to the plot\n",
    "plt.legend(handles=[rmip_handle, eqtl_handle], loc='upper right', bbox_to_anchor=(1, 1), frameon=False)\n",
    "\n",
    "# Save the plot to Google Drive\n",
    "svg_file_path = 'RMIP_and_eQTL_hits_with_specific_genes.svg'\n",
    "plt.savefig(svg_file_path, format='svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50f18d25-84d4-4873-b1ae-f01b4f989a82",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
